### PR TITLE
Add Tiny Well-Known Binary decoder and encoder.

### PR DESCRIPTION
**Pull request by DoctorLoki**
## Description

Add Tiny Well-Known Binary decoding and encoding capability with unit tests.

## Check List

Have you:

- Added unit tests? Yes

- Add cmprefimpl tests? No

- Updated release notes? No

## Related Issue

Implements https://github.com/peterstace/simplefeatures/issues/54

### CODE REVIEW

File: geom/twkb_parser.go
Comment (DoctorLoki) [inline geom/twkb_parser.go:313]: This part is of concern to me.
See multiple polygon examples at https://github.com/TWKB/twkb.js/blob/master/test/twkb.spec.js#L43 which seem to do things differently, and particularly https://github.com/TWKB/twkb.js/blob/master/test/twkb.spec.js#L169 but compare that with the spec and you can see why I'm concerned.

File: geom/twkb_parser.go
Comment (DoctorLoki) [inline geom/twkb_parser.go:178]: The spec seems unclear about whether any further data is needed if the metadata header byte indicated the object is empty. In my tests I have assumed no extra bytes are required. This may b at variance from the reference implementation https://github.com/TWKB/twkb.js/blob/master/src/readBuffer.js which sets `is_empty` but then ignores the field. So I may need to change the tests and implementation to also require a zero number of points/rings/etc to be present.

Comment (DoctorLoki) [reply geom/twkb_parser.go:178]: Update: PostGIS seems to agree with my implementation:
```
=> select ST_AsTWKB('POINT EMPTY'::geometry);
 st_astwkb 
-----------
 \x0110

surveys=> select ST_AsTWKB('POLYGON EMPTY'::geometry);
 st_astwkb 
-----------
 \x0310
```

File: geom/twkb_parser.go
Comment (DoctorLoki) [reply geom/twkb_parser.go:313]: It seems that at least one implementation emits closed rings when encoding (note there are 5 points given for each ring):
```
=> select ST_AsTWKB('POLYGON((0 0,2 0,2 2,0 2,0 0),(0 0,0 1,1 1,1 0,0 0))'::geometry);
                      st_astwkb                       
------------------------------------------------------
 \x03000205000004000004030000030500000002020000010100
```
When decoding, both closed and unclosed rings are accepted and correctly handled as below:
```
=> select ST_AsText(ST_GeomFromTWKB(E'\\x03000205000004000004030000030500000002020000010100'));
                      st_astext                       
------------------------------------------------------
 POLYGON((0 0,2 0,2 2,0 2,0 0),(0 0,0 1,1 1,1 0,0 0))

=> select ST_AsText(ST_GeomFromTWKB(E'\\x030002040000040000040300040003000202000001'));
                      st_astext                       
------------------------------------------------------
 POLYGON((0 0,2 0,2 2,0 2,0 0),(0 0,0 1,1 1,1 0,0 0))
```
So I think the parser should handle both closed and unclosed rings.

Comment (DoctorLoki) [reply geom/twkb_parser.go]: I've now implemented that change, to allow the parser to accept either closed or unclosed rings.

File: geom/twkb_parser.go
Comment (albertteoh) [inline geom/twkb_parser.go:520]: Did you mean to use `binary.Varint`?

File: geom/twkb_parser.go
Comment (albertteoh) [inline geom/twkb_parser.go:323]: What's the reason for `0+d` instead of just `d`?

File: geom/twkb_parser.go
Comment (albertteoh) [inline geom/twkb_parser.go:522]: The [docs](https://pkg.go.dev/encoding/binary#Varint) differentiate `n == 0` and `n < 0`, which have different meanings. Do you think it's worth conveying these meanings in separate error messages?

```
n == 0: buf too small
n  < 0: value larger than 64 bits (overflow)
        and -n is the number of bytes read
```

File: geom/twkb_parser.go
Comment (albertteoh) [inline geom/twkb_parser.go:184]:
```suggestion
		return p.parserError("cannot unmarshal: lacking extended precision byte")
```

Comment (albertteoh) [reply geom/twkb_parser.go:184]: nit: I suggest adding a `:` after each `cannot unmarshal` in error messages (same below) as I feel we're annotating two error messages here and it helps with reading the error message:
- The fact that unmarshalling failed.
- The reason why unmarshalling failed.

File: geom/twkb_parser.go
Comment (albertteoh) [inline geom/twkb_parser.go:427]: I noticed that `annotateError` is used in some places but not others. 

Would it make sense to annotate the error once in `UnmarshalTWKB` (bottom of the call stack) and instead of `p.parserError`, return the built-in error via `fmt.Errorf(...)` or `errors.New(...)` instead? It would help reduce potential duplication of the byte position annotation in error messages.

File: geom/twkb_parser.go
Comment (DoctorLoki) [reply geom/twkb_parser.go:184]: Good point, I'll make the error messages a bit clearer.

File: geom/twkb_parser.go
Comment (DoctorLoki) [reply geom/twkb_parser.go:323]: It was just to reinforce the fact we're reading from the first point (at offset 0). It's obviously not strictly necessary. I can remove it.

File: geom/twkb_parser.go
Comment (DoctorLoki) [reply geom/twkb_parser.go:427]: I was trying to use it only in return paths where the byte position wasn't otherwise being reported. Good point about annotating in one spot, I'll try to tidy that up.

File: geom/twkb_parser.go
Comment (DoctorLoki) [reply geom/twkb_parser.go:520]: Good point, `binary.Varint` reverses the zig-zag encoding so I don't need to do it myself. Will change.

File: geom/twkb_parser.go
Comment (albertteoh) [reply geom/twkb_parser.go:323]: Okay, that makes sense, we can leave it there, or maybe put it in an `offset` variable if you feel it helps with readability. No strong opinion either way.

File: geom/twkb_parser.go
Comment (DoctorLoki) [reply geom/twkb_parser.go:522]: I didn't bother differentiating those because they're going to lead to the parse being aborted either way, but you're right, separate error messages could help with debugging.

File: geom/twkb_parser.go
Comment (albertteoh) [inline geom/twkb_parser.go:65]: The `%w` verb allows errors to be used with `errors.Is` and `errors.As`.

See: https://go.dev/blog/go1.13-errors

```suggestion
	return fmt.Errorf("TWKB parsing error at byte %d: %w", p.pos, err)
```

Comment (albertteoh) [reply geom/twkb_parser.go:65]: I noticed a few more cases in this PR where we could use the `%w` verb as well.

Comment (DoctorLoki) [reply geom/twkb_parser.go]: Nice, I'm using the `%w` verb now.

File: geom/twkb.go
Comment (albertteoh) [inline geom/twkb.go:7]: nit: These constants could technically be successfully interchanged with one another. That is, usages of `twkbTypeMultiPoint ` could be accidentally but successfully replaced with `twkbMaxDimensions` or `twkbHasIDs` and succeed. Maybe we can consider typing them?

```
type geomKinds int
type metaHeaders int
```

**Review by albertteoh — APPROVED** [review]
LGTM. Added a non-blocking suggestion. I might have to defer to @peterstace re: the GIS-specific logic.

File: geom/twkb.go
Comment (DoctorLoki) [reply geom/twkb.go]: I've introduced types for the geometry kinds and metadata headers. You're right, I think it's cleaner and more comprehensible this way.

**Comment by DoctorLoki** [comment]
Note to self: TODO:
* Verify every test case using PostGIS or equivalent, for decoding, and whatever the valid subset is for encoding.
* Determine if we also need methods like `(p Point) AsTWKB()` etc.

File: geom/twkb_write.go
Comment (albertteoh) [inline geom/twkb_write.go:492]: Just checking: we're certain this should _never_ happen?

File: geom/twkb_write.go
Comment (albertteoh) [inline geom/twkb_write.go:486]: Curious, why 2?

File: geom/twkb_write.go
Comment (albertteoh) [inline geom/twkb_write.go:513]: nit: perhaps the more idiomatic approach is `for _, id := range w.idList { w.writeSignedVarint(id) }`?

File: geom/twkb_write.go
Comment (albertteoh) [inline geom/twkb_write.go:371]: Just checking: did you mean to place this comment here?

File: geom/twkb_write.go
Comment (albertteoh) [inline geom/twkb_write.go:335]: It seems (I could be wrong) that each geom's write method has a common block of code to start with:

```
	w.writeTypeAndPrecision(<the type enum>)

	if ctype := gc.CoordinatesType(); ctype != w.ctype {
		return fmt.Errorf("mismatched coordinate dimensions got %s expected %s", ctype, w.ctype)
	}

	if gc.IsEmpty() {
		w.writeIsEmptyHeader()
		return nil
	}
	w.writeInitialHeaders()
```

Perhaps this could be extracted out into a method?

File: geom/twkb_write.go
Comment (albertteoh) [inline geom/twkb_write.go:99]:
```suggestion
	w.hasExt = hasZ || hasM
```

File: geom/twkb_write.go
Comment (albertteoh) [inline geom/twkb_write.go:448]: Maybe I'm missing something but why is there a need to check `if !w.bboxValid`?

File: geom/twkb_parser.go
Comment (DoctorLoki) [reply geom/twkb_parser.go:32]: I've made `TWKBParser` and `TWKBWriter` unexported now.

File: geom/twkb_write.go
Comment (DoctorLoki) [reply geom/twkb_write.go:492]: Yes, it'd be a programming error if we're trying to write additional headers before the initial headers have been written. I've checked all the cases, so this shouldn't ever happen, which is why I've written it as a `panic`.

File: geom/twkb_write.go
Comment (DoctorLoki) [reply geom/twkb_write.go:371]: That comment at the end of the function was to explain why there's nothing else to be done in that function (because it's an empty object so we stop right there). I'll rewrite that to make it clearer.

File: geom/twkb_write.go
Comment (DoctorLoki) [reply geom/twkb_write.go:335]: Although it's true there's common(ish) code at the top of each geometry type's write function, the behaviour of `IsEmpty()` for the various types may mean we want to treat them differently. I'll do some more testing before considering amalgamating any of this code.

File: geom/twkb_write.go
Comment (DoctorLoki) [reply geom/twkb_write.go:99]: In general that transformed code is not identical (but I think it is valid in this case). I'll make the change.

File: geom/twkb_write.go
Comment (DoctorLoki) [reply geom/twkb_write.go:448]: Yep, you're right, that was a remnant of some refactored code. It was originally inside the inner loop within an `if` statement having the same condition, but that was incorrect because it would set the `bboxValid` flag to true after the X co-ordinate was examined, rather than after the entire first point was examined (which had a weird side-effect of making the bounding box min and max extents wrong for all dimensions except for X).

I'll add a comment to explain why the bbox becomes valid after the first point, i.e. why the code is where it is.

File: geom/twkb_write.go
Comment (DoctorLoki) [reply geom/twkb_write.go]: I've now separated the TWKB writer's internal buffer into three separate buffers, and I think it does clarify things, so I've committed that version.

**Comment by DoctorLoki** [comment]
Current status of this code:
* I have verified that every test case matches how PostGIS functions. It is only the `IsEmpty()` cases that may need to be adjusted. See discussion earlier.
* I haven't yet implemented methods such as `(p Point) AsTWKB()` but it would be straightforward to do so.

**Comment by peterstace** [comment]
Thanks for putting so much work into this! I should get an opportunity to take a deep look either this weekend on early next week 😄

File: geom/twkb_parser.go
Comment (DoctorLoki) [reply geom/twkb_parser.go]: Update: I'm fairly convinced that the currently implemented approach is correct. That is, if a polygon ring's final point is the same as its first point, the TWKB decoder silently allow that, but also allow the case that the final point isn't the same as its first point (in which case close the ring by appending the first point as its final point). This seems to be what PostGIS does.

My implementation also allows both forms to be generated from the TWKB encoder, with a Boolean flag to control that. Obviously, for size of output it's generally better to omit the final point of a ring as it is redundant information. This seems not to be what PostGIS does, it seems to always output the final point, and there doesn't seem to be a way to control this behaviour. This simply means my implementation is a little more versatile, which is not a problem.

File: geom/twkb_parser.go
Comment (peterstace) [inline geom/twkb_parser.go:335]: Very minor nit: could do `coords = append(coords, coords[:p.dimensions]...)` instead of the loop.

File: geom/twkb_parser.go
Comment (peterstace) [inline geom/twkb_parser.go:482]: The conversion from the `p.refpoint[d]` value to `float64` is technically lossy above 2^52, since floats start "skipping" integer values above that point. I don't really see an easy way around it though... (e.g. we probably _don't_ want to mess around with arbitrary precision arithmetic). I don't think it will be a problem in practice, but thought I'd bring it up anyway in case you see an easy solution.

**Review by peterstace — COMMENTED** [review]
I've reviewed the parsing only so far (I'll get to the writer soon!). Everything is looking pretty good so far!

File: geom/twkb.go
Comment (DoctorLoki) [reply geom/twkb.go:32]: Sure, the ZigZag code can be unexported, it was exported really only for testing.

File: geom/twkb_parser.go
Comment (DoctorLoki) [reply geom/twkb_parser.go:24]: Your understanding of the bounding box situation is correct in all ways. I'll add the suggested cheap function for just reading the bounding box header if it exists in the data stream.

File: geom/twkb_write.go
Comment (peterstace) [inline geom/twkb_write.go:10]: Regarding the function signature, a few things stick out to me:

- The user shouldn't have to supply `hasZ` and `hasM`. Instead, `MarshalTWKB` should be able to get this info using `geom.CoordinatesType().Is3D()` and `geom.CoordinatesType().IsMeasured()`.
- `precXY` should always be specified, but `precZ` and `precM` only make sense for 3D and measured geometries.
- My understanding of the spec is that writers _shouldn't_ close rings. So I don't think we need to make that behaviour configurable. Does that match your understanding?
- `hasSize` and `hasBBox` seem sensible to allow to be configured, although IMO ideally we would default these to both `false`.
- `idList` is sensible to allow, some (many?) users wouldn't need it.

What do you think about using the functional options pattern? The signature could look something like this:

```
func MarshalTWKB(g Geometry, precXY int, opts ...TWKBWriterOption) ([]byte, error) {
    var os twkbWriterOptionSet
    os.precZ = precXY // default if not set by opt
    os.precY = precXY // default if not set by opt
    for _, opt := range opts {
        opt(&os)
    }
    ... // the rest of the logic
}

type TWKBWriterOption func(*twkbWriterOptionSet)

type twkbWriterOptionSet struct {
    precZ, precM int
    hasSize, hasBBox bool
    idList []int64
}

func TWKBPrecisionZ(int precZ) TWKBWriterOption {
    return func(s *twkbWriterOptionSet) {
        s.precZ = precZ
    }
}

... // the rest of the options

```

File: geom/twkb_write.go
Comment (peterstace) [inline geom/twkb_write.go:133]: Nit: since you know the total number of bytes needed (the sum of the header, bbox, and contents), you could pre-allocate the data slice.

**Review by peterstace — CHANGES_REQUESTED** [review]
Overall, looks good! A few things to discuss though.

File: geom/twkb_parser.go
Comment (DoctorLoki) [reply geom/twkb_parser.go]: Concerning the bounding box header parsing:

I've modified the function signature for
```UnmarshalTWKBWithHeaders(twkb []byte, opts ...ConstructorOption) (g Geometry, bbox []Point, ids []int64, err error)```

Previously that function was returning bounding box information as integers (which exposed implementation internals, naughty!) but now the bounding box is returned as a slice of `Point`. If the slice is empty, there was no bounding box header. Otherwise it contains a minimum and a maximum `Point`. Because `Point` is used, the X, Y, Z and M coordinate values may all be present; no data loss occurs.

I've also added two new functions:
```UnmarshalTWKBBoundingBoxHeader(twkb []byte) (bbox []Point, err error)```
and
```UnmarshalTWKBEnvelope(twkb []byte) (bbox Envelope, hasBBox bool, err error)```

The former function only parses the bounding box header, so it returns without parsing the full geometry.
The semantics of the `Point` slice are the same as discussed above.

The latter function also only parses the bounding box header, but returns it as an `Envelope`.
The `hasBBox` flag is false if there's no such header. Otherwise, the envelope is valid.
Due to the definition of `Envelope` only X and Y values can be reported this way.
(Any Z and M values are omitted because they cannot be represented in an `Envelope`.)

I've added tests checking all these functions behave correctly.

File: geom/twkb_parser.go
Comment (DoctorLoki) [reply geom/twkb_parser.go:86]: Regarding the information redundancy between `ctype` and `hasZ` and `hasM` and `dimensions` within the TWKB parser:

Although yes there's redundancy of information, the way these values are used differs.
* `hasZ` and `hasM` are derived from the TWKB headers, and thus indicate what was present in the input bytes.
* `dimensions` is then deduced from `hasZ` and `hasM` and controls how coordinates are parsed.
* `ctype` is also deduced from them, and is only used when constructing a `Point` object.

I feel there's a separation here, where `hasZ` and `hasM` and `dimensions` are TWKB-specific decoding state variables, whereas `ctype` is part of the interface for constructing geometries. So to me the separation of those two kinds of variables serves a purpose. The former variables act as documentation which can be read alongside the spec to verify the implementation, the latter `ctype` variable is just a convenience for creating geometries. Shifting all those status flags into the one `ctype` variable I feel might obscure the code a bit.

That being said, I'm happy to combine them if you feel strongly there should be no information redundancy.

File: geom/twkb_parser.go
Comment (DoctorLoki) [reply geom/twkb_parser.go:600]: Regarding converting coordinates to/from `float64`:

There are four bits in the TWKB header to indicate a signed integer for the power of 10 scaling factor for X and Y coordinates. The case we're probably most interested in is where we are keeping 7 decimal places in a latitude or longitude, which is the maximum number of decimal places TWKB lets us retain. Now 10^7 is approximately 2^24, and if we're keeping whole numbers up to +/-180 in front of that, we can assume another 8 bits on top, so it probably amounts to 32 bits of significant information for most real world cases.

My _feeling_ is the `int64` data we're dealing with won't have float conversion roundoffs, and probably won't lose much precision even after scaling. Besides, I don't see a good way around a hypothetical float roundoff, it's simply inherent to the efficiency of using floats.

File: geom/twkb_write.go
Comment (DoctorLoki) [reply geom/twkb_write.go:10]: Regarding your comments about the encoder:
* Yes, I'll rework the function to allow options. I'd considered it, but figured I'd wait to see what approach you'd prefer.
* Agreed that many users won't use the ID list. It's there for adherence to the spec.
* Agreed that many flags can often be omitted or defaulted to false.
* Technically, supplying `hasZ` and `hasM` separately from the `Geometry` could provide additional functionality of stripping those coordinate values during encoding if they are `false` (although that's not the way the code is currently written). If you're happy to always deduce these from the geometry, those parameters can be omitted as you mentioned.
* You're correct that the spec says the encoder shouldn't close rings. The reason I added that capability was to allow byte-for-byte comparison against an actual PostGIS implementation which does seem to close rings. If possible I'd like to retain the capability, but hide it behind an option.

Comment (peterstace) [reply geom/twkb_write.go]: Sounds good!

File: geom/twkb_parser.go
Comment (peterstace) [inline geom/twkb_parser.go:90]: The `Envelope` can represent an empty Envelope, so the bool return isn't really needed. Should probably also use the regular envelope constructor rather than the unchecked variant (I don't see any reason not to).

```suggestion
func UnmarshalTWKBEnvelope(twkb []byte) (Envelope, error) {
	p := newtwkbParser(twkb)
	if err := p.parseHeaders(); err != nil {
		return Envelope{}, p.annotateError(err)
	}
	if !p.hasBBox {
		return Envelope{}, nil
	}
	return NewEnvelope([]XY{
		XY{
			p.scalings[0] * float64(p.bbox[0]),
			p.scalings[1] * float64(p.bbox[2]),
		},
		XY{
			p.scalings[0] * float64(p.bbox[0]+p.bbox[1]),
			p.scalings[1] * float64(p.bbox[2]+p.bbox[3]),
		},
	})
}
```

File: geom/twkb_parser.go
Comment (peterstace) [reply geom/twkb_parser.go:100]: What you say makes senes, let's leave it as is.

**Review by peterstace — COMMENTED** [review]
Looking really good! Just a new comment about Envelope being able to be empty, but once that's fixed we should be able to merge 🍰

File: geom/twkb_parser.go
Comment (DoctorLoki) [reply geom/twkb_parser.go]: I've modified the envelope function as you suggest, fixing the tests and also adjusting the comment to clarify.

**Review by peterstace — APPROVED** [review]

**Comment by peterstace** [comment]
Thank you for contributing this large and significant feature 😁 🍰

**Pull request merged**
